docs/r1cs-docs-example.md

```diff
@@ -0,0 +1,530 @@
+The shuffle `verify` function verifies that, given a shuffle proof and a list of committed inputs and outputs, the outputs are a valid reordering of the inputs.
+
+Because only the prover knows the scalar values of the inputs and outputs, and the verifier only sees the (blinded) committed inputs and outputs, the verifier has no knowledge of what the underlying inputs and output values are. Therefore, the only information the verifier learns from this protocol is whether or not the committed outputs are a valid shuffle of the committed inputs - this is why it is a zero-knowledge proof.
+
+```rust
+extern crate bulletproofs;
+extern crate curve25519_dalek;
+extern crate merlin;
+extern crate rand;
+
+use bulletproofs::r1cs::*;
+use bulletproofs::{BulletproofGens, PedersenGens};
+use curve25519_dalek::ristretto::CompressedRistretto;
+use curve25519_dalek::scalar::Scalar;
+use merlin::Transcript;
+use rand::thread_rng;
+
+// Shuffle gadget (documented in markdown file)
+
+/// A proof-of-shuffle.
+struct ShuffleProof(R1CSProof);
+
+impl ShuffleProof {
+    fn gadget<CS: ConstraintSystem>(cs: &mut CS, x: &[Variable], y: &[Variable]) {
+        let z = cs.challenge_scalar(b"shuffle challenge");
+
+        assert_eq!(x.len(), y.len());
+        let k = x.len();
+
+        if k == 1 {
+            cs.constrain(y[0] - x[0]);
+            return;
+        }
+
+        // Make last x multiplier for i = k-1 and k-2
+        let (_, _, last_mulx_out) = cs.multiply(x[k - 1] - z, x[k - 2] - z);
+
+        // Make multipliers for x from i == [0, k-3]
+        let first_mulx_out = (0..k - 2).rev().fold(last_mulx_out, |prev_out, i| {
+            let (_, _, o) = cs.multiply(prev_out.into(), x[i] - z);
+            o
+        });
+
+        // Make last y multiplier for i = k-1 and k-2
+        let (_, _, last_muly_out) = cs.multiply(y[k - 1] - z, y[k - 2] - z);
+
+        // Make multipliers for y from i == [0, k-3]
+        let first_muly_out = (0..k - 2).rev().fold(last_muly_out, |prev_out, i| {
+            let (_, _, o) = cs.multiply(prev_out.into(), y[i] - z);
+            o
+        });
+
+        // Constrain last x mul output and last y mul output to be equal
+        cs.constrain(first_mulx_out - first_muly_out);
+    }
+}
+```
+
+In this example, `ShuffleProof::gadget()` is private function that adds constraints to the constraint system that enforce that `y` (the outputs) are a valid reordering of `x` (the inputs). 
+
+First, the function gets a challenge scalar `z` by calling the `ConstraintSystem::challenge_scalar`. This challenge is generated from commitments to high-level variables that were passed to the `ConstraintSystem` when it was created. As noted in the `challenge_scalar` documentation, it is important to ensure that the challenge is sound by making sure that low-level variables are bound to the high-level variables. 
```

Comment:
that makes sense. changed to:

As noted in the `challenge_scalar` documentation, making sure that the challenge circuit is sound requires analysis. In this example, the challenge circuit is sound because the challenge is bound to all of the shuffle inputs and outputs, since the inputs and outputs are high-level variables.